File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-5.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0;
        int r = nums.size() - 1;

        while (l <= r) {
            int mid = (r + l) / 2;
            cout << mid << endl;

            if (nums[mid] == target) {
                return mid;
            }
            if (nums[mid] >= nums[l]) {
                if (target > nums[mid] || target < nums[l]) {
                    l = mid + 1;
                }
                else {
                    r = mid - 1;
                }
            }
            else {
                if (target < nums[mid] || target > nums[r]) {
                    r = mid - 1;
                }
                else {
                    l = mid + 1;
                }

            }

        }
        return -1;
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-5.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        memo = {}

        def dfs(last, i):
            if i >= len(nums):
                return 0
            if (last, i) not in memo:
                if last == -1 or nums[last] < nums[i]:
                    memo[(last, i)] = max(1 + dfs(i, i + 1), dfs(last, i + 1))
                else:
                    memo[(last, i)] = dfs(last, i + 1)

            return memo[(last, i)]
        
        return dfs(-1, 0)